File: Aula_Ricardo/Programação_orientada_a_objetos/python/Erros_personalizados.py
class SaldoInsuficente(Exception):
    def __init__(self, message="Saldo insuficiente"):
        self.message = message
        super().__init__(self.message)

class Conta_bancaria:
    def __init__(self, saldo=0):
        self.saldo = saldo

    def depositar(self, quantia):
        self.saldo += quantia

    def sacar(self, quantia):
        if quantia > self.saldo:
            raise SaldoInsuficente()
        self.saldo -= quantia

conta = Conta_bancaria(500)
try: 
    conta.sacar(499)
except SaldoInsuficente as e:
    print(e)

File: Aula_Ricardo/Programação_orientada_a_objetos/go/Erros_personalizados.go
package main

import (
	"fmt"
)

type SaldoInsuficiente struct {
	Message string
}

func (e *SaldoInsuficiente) Error() string {
	return e.Message
}

type ContaBancaria struct {
	Saldo float64
}

func (c *ContaBancaria) Depositar(quantia float64) {
	c.Saldo += quantia
}

func (c *ContaBancaria) Sacar(quantia float64) error {
	if quantia > c.Saldo {
		return &SaldoInsuficiente{Message: "Saldo insuficiente"}
	}
	c.Saldo -= quantia
	return nil
}

func main() {
	conta := &ContaBancaria{Saldo: 500}

	err := conta.Sacar(499)
	if err != nil {
		fmt.Println(err)
	} else {
		fmt.Println("Saque realizado com sucesso!")
	}
}
